be/src/storage/binlog_manager.cpp

```diff
@@ -43,9 +43,195 @@ BinlogManager::~BinlogManager() {
+    BinlogFileMetaPBPtr last_file_meta;
+    if (!recovered_file_metas.empty()) {
+        auto& file_meta = recovered_file_metas.back();
+        BinlogLsn start_lsn(file_meta->start_version(), file_meta->start_seq_id());
+        if (start_lsn <= min_lsn) {
+            // the binlog for the version is only in one file,
+            // and finish to recover the version
+            return Status::OK();
+        } else if (start_lsn.version() == version) {
+            // the last recovered file only contains part of binlog
+            last_file_meta = file_meta;
+        } else {
+            // the last file is just the start of the last recovered version
+            DCHECK(start_lsn.seq_id() == 0);
+        }
+    }
+
+    bool find_all_binlog = false;
+    while (file_id_it != file_ids.rend()) {
+        int64_t file_id = *file_id_it;
+        file_id_it++;
+        StatusOr<BinlogFileMetaPBPtr> status_or =
+                _recover_file_meta_for_version(version, file_id, last_file_meta.get());
+        if (!status_or.ok() && !status_or.status().is_not_found()) {
+            return status_or.status();
+        }
+        if (status_or.status().is_not_found()) {
+            useless_file_ids->push_back(file_id);
+            continue;
+        }
+        BinlogFileMetaPBPtr file_meta = status_or.value();
+        recovered_file_metas.push_back(file_meta);
+        BinlogLsn lsn(file_meta->start_version(), file_meta->start_seq_id());
+        last_file_meta = file_meta;
+        if (lsn <= min_lsn) {
+            find_all_binlog = true;
+            break;
+        }
+    }
+
+    if (!find_all_binlog) {
+        std::string errMsg = fmt::format("Failed to recover version {} because of incomplete binlog, tablet {}",
+                                         version, _tablet_id);
+        std::string meta_msg =
+                last_file_meta == nullptr ? "null" : BinlogUtil::file_meta_to_string(last_file_meta.get());
+        LOG(ERROR) << errMsg << ", last file meta: " << meta_msg;
+        return Status::InternalError(errMsg);
+    }
+
+    return Status::OK();
+}
+
+StatusOr<BinlogFileMetaPBPtr> BinlogManager::_recover_file_meta_for_version(int64_t version, int64_t file_id,
+                                                                            BinlogFileMetaPB* last_file_meta) {
+    BinlogLsn lsn_upper_bound;
+    if (last_file_meta == nullptr) {
+        lsn_upper_bound = BinlogLsn(version + 1, 0);
+    } else {
+        lsn_upper_bound = BinlogLsn(last_file_meta->start_version(), last_file_meta->start_seq_id());
+    }
+    std::string file_path = BinlogUtil::binlog_file_path(_path, file_id);
+    StatusOr status_or = BinlogFileReader::load_meta(file_id, file_path, lsn_upper_bound);
+    if (!status_or.ok() && !status_or.status().is_not_found()) {
+        std::string errMsg = fmt::format("Failed to recover version {} because of load error, tablet {}, file_path: {}",
```

Comment:
all `errMsg` should be converted to `err_msg`

Reply:
fixed